Index: libchewing/src/porting_layer/include/sys/plat_posix.h
===================================================================
--- libchewing.orig/src/porting_layer/include/sys/plat_posix.h	2012-03-26 22:48:27.021991254 +0800
+++ libchewing/src/porting_layer/include/sys/plat_posix.h	2012-04-03 17:06:14.352127987 +0800
@@ -20,6 +20,11 @@
 #define PLAT_UNLINK(path) \
 	unlink(path)
 
+//GNU Hurd doesn't define PATH_MAX
+#ifndef PATH_MAX
+#define PATH_MAX 4096
+#endif
+
 #ifdef __cplusplus
 extern "C"
 {
